subprojects/internal-testing/src/main/groovy/org/gradle/util/FlakyTest.groovy

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.util
+
+
+import org.gradle.util.internal.AlwaysTrue
+
+import java.lang.annotation.ElementType
+import java.lang.annotation.Retention
+import java.lang.annotation.RetentionPolicy
+import java.lang.annotation.Target
+
+/**
+ * Marks a test as flaky. The flaky tests are quarantined to run
+ * at a special stage so it won't interrupt normal build pipeline.
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target([ElementType.TYPE, ElementType.METHOD])
+@interface FlakyTest {
+    /**
+     * If the closure evaluated to true, it will be ignored
+     * in normal build pipeline, but will be executed in flaky
+     * quarantine build. Otherwise, it's treated as a normal test.
+     */
+    Class<? extends Closure> flakyIf() default AlwaysTrue.class;
+
+    /**
+     * Optional description for the issue, like a issue tracker URL.
+     * @return
+     */
+    String issue() default "";
```

Comment:
I wonder if the issue should be necessary. Or if there should be a `reason`/`because` obligatory text, which may be a link to the issue.